### ISSUE TITLE
SELECT target FROM Relations - does not work in HTML Report

### ISSUE THREAD

**Issue by fokkerispens** [archimatetool/archi#400]
### Version of Archi, Operating System
4.3 on Windows 10. Tested on browsers Chrome and Edge and inside Archi (HTML Preview)

### Expected Behaviour
A list of id's should appear, when executing "SELECT target FROM Relations" in the query tab of an HTML report. E.g. "SELECT source FROM Relations" or "SELECT * FROM Relations" do return a list of id's. Selecting target doesn't

### Actual Behaviour
Nothing happens, the input returns to empty

### Steps to Reproduce the Behaviour
1. Generate of preview HTML report
2. Go to QUERY Tab
3. Execute "SELECT target FROM Relations"

**Comment by jbsarrodie** [comment]
Hi. That's because 'target' is a reserved word in alasql engine. I'll change it to something else (maybe 'tgt') in next release.

**Comment by fokkerispens** [comment]
Hi, thanx for the fast response! Do you happen to know a work-around. Because now I can't join on Relations.target = Elements.id and can't generate a cross reference.

**Comment by Phillipus** [comment]
You could edit the frame.stg file found in:

plugins/com.archimatetool.reports_4.3.0.201809101333/templates/st

Line 46:

```
dataRelations.push({id:"^c.Id^", type:"^c.Class.SimpleName^", name:decodeURL("^c.Name;format="url-encode"^"), description:decodeURL("^c.Description;format="url-encode"^"), source:"^c.Source.Id^", target:"^c.Target.Id^"\});
```

Change target to tgt.

**Comment by fokkerispens** [comment]
Thanx! I will change 'target' to 'targetid' and 'source' to 'sourceid'. So the naming will be more in line with the meaning. An alternative could be 'target-id' and 'source-id'.

Could you please let me know which names you will use. Because I want to store my queries and don't want to rewrite them :-)

BTW: Storing queries in Archi or using predefined queries in HTML report. would be nice feature of version 4.4 :-)

**Comment by jbsarrodie** [comment]
I'd go for "sourceid" and "targetid", but I'd like to also check other columns (at least for the naming, like in properties that should refer to "conceptid")

@Phillipus I'll do it during the week-end and will check other columns

**Comment by fokkerispens** [comment]
Thanx! 

Will you let me now which columns you change and their names, so I can change them in my version of frame.stg an re-use my queries in the next version of Archi

**Comment by fokkerispens** [comment]
Hi,

I changed target in targetid and and source in sourceid in the frame.stg file found in: plugins/com.archimatetool.reports_4.3.0.201809101333/templates/st on line 46.

And I got the following SQL statement working:

select S.name as sourcename, R.type, T.name as targetname from Relations R, Elements S, Elements T where S.id = R.sourceid and T.id = R.targetid

Which gives a cross reference between elements and the kind of relation between them.

Four snags: 
1. Use [table].[attribute] as syntax for name
2. Use old fashioned where statement, I didn't get the "join"keyword working
3. Watch the use of uppercase and lowercase. They should match precisely
4. Depending on your model: wait, because it takes some time. I tested on a small model and it took me four seconds.

If you change any other columns, I please let me know.

Thanx again!

**Comment by jbsarrodie** [comment]
Hi,

>Use [table].[attribute] as syntax for name
>Depending on your model: wait, because it takes some time. I tested on a small model and it took me four seconds.

2 and 4 are linked : not relaying on "join" keywords leads to very (very!) long execution time. I once converted a query that was taking more than 2min to load... once changed it was less than 5sec.

>Watch the use of uppercase and lowercase. They should match precisely

Yes, maybe it would make sens to revert to lowercase everywhere, or set the `alasql.options.casesensitive` flag to `false`.

**Comment by fokkerispens** [comment]
Hi,

That makes sense, no query optimizing like SQL server does (which I'm used to :-). 

After some research I found out I can save the query results into CSV with:

SELECT * INTO CSV("ArchimateRelations.csv",{headers:true}) FROM Relations;
SELECT * INTO CSV("ArchimateElements.csv",{headers:true}) FROM Elements;
SELECT * INTO CSV("ArchimateProperties.csv",{headers:true}) FROM Properties;

I think I will do that en read the results into a database/query tool in which I can store my predefined queries en views. Of cause that's not 'real-time' and takes some manual steps, but I need it to facilitate the governance of our architectural models. So I don't need 'real-time' result. But I can imagine a lot of use cases which do deed a more 'real-time' and user friendly approach. Querying the architectural model  (and storing predefined queries) is a very powerful feature, so please continue developing it. 

By the way: 
(1) Since Properties.id is a primary key, I think the name should stay 'id' in the frame.stg file
(2) I like the alasql.options.casesensitive flag to false option, but have no I idea of where to set it

Thanx again!

**Comment by jbsarrodie** [comment]
>After some research I found out I can save the query results into CSV with:

Yes, but in this case you'd better simply use the CSV export from Archi.

>I think I will do that en read the results into a database/query tool in which I can store my predefined queries en views. [...] Querying the architectural model (and storing predefined queries) is a very powerful feature, so please continue developing it.

The next version (full rewrite) of the HTML report will support "saving" queries in the model itself (through some model-level properties or object, yet to be defined) and will also allows you to passe the query through the URL, so bookmarking a query will be possible too.

>Since Properties.id is a primary key, I think the name should stay 'id' in the frame.stg file

But Properties.id is **not** a primary key: this is the key of the object that has the property set.

>I like the alasql.options.casesensitive flag to false option, but have no I idea of where to set it

You can set it in `plugins/com.archimatetool.reports_4.3.0.201809101333/templates/st/frame.stg` around line 211:
````
[...]
alasql.options.logtarget = 'myconsole';
alasql.options.casesensitive = false;
alasql.prompt('myprompt','useid');
[...]
````

**Comment by aranxas** [comment]
Dear all,

the problem with 

`SELECT target FROM Relations`

seems to still exist in 4.3.2.  A potentially related problem is that

`SELECT value FROM Properties`

does not yield results either. Or rather the recognized SQL command in this case is

`SELECT COLUMN _ FROM Properties`

and it outputs the whole Properties table.

**Comment by aranxas** [comment]
Also if you run 

`SELECT * FROM Elements`

the column 'description' is always empty, irrespective of whether the Documentation in the ArchiMate model is actually populated.

**Comment by jbsarrodie** [comment]
@Phillipus I'll look at this. This is partly related to another issue about Alasql reserved words which should be fixed, and some informations are still not exported. I'll provide a fix for next release ;-)

**Comment by aranxas** [comment]
I have played around a bit with the alasql prompt. I also added a table
with the folders locally as I want to run queries for components in
specific folders.

What would be a nifty has is the following:

If you have a diagram with some graphical element (e.g. a note) on it that
can have properties it would be cool to double-click the note and have
alasql run a query stored in the properties of the element as a simple
reporting feature.

Currently I keep my queries in a text editor but having them embedded in
the html report would really be nice.

Am Do., 7. Feb. 2019, 11:26 hat Jean-Baptiste Sarrodie <
notifications@github.com> geschrieben:

**Comment by jbsarrodie** [comment]
This is fixed in the new **html-query-fix** branch.

Here are the changes:
* Table **Relations** has been renamed to **Relationships**
* **Relations.target** has been renamed to **Relationships.targetid**
* **Relations.source** has been renamed to **Relationships.sourceid**
* **Properties.id** has been renamed to **Properties.conceptid**
* **Properties.key** has been renamed to **Properties.propkey**
* **Properties.value** has been renamed to **Properties.propvalue**
* View are no more exported (they were exported as elements, which they are not)

**Comment by Phillipus** [comment]
Fixed in Archi 4.3.3.
